src/main/java/org/jahia/modules/appshell/Main.java

```diff
@@ -144,49 +249,145 @@ private void setCustomAttributes(JahiaUser currentUser, HttpServletRequestWrappe
         wrapper.setAttribute("config", config.toString());
     }
 
+    public void updateAppInfos() {
+        List<Bundle> packages = getBundlesWithPackages();
 
-    public List<String> getApplicationScripts(String appName, String key) throws JSONException, IOException {
-        LinkedList<String> resources = new LinkedList<>();
+        Map<String,AppInfo> newAppInfos = new LinkedHashMap<>();
+        for (Bundle bundle : packages) {
+            updateBundleInfo(newAppInfos, bundle);
+        }
+        appInfos = newAppInfos;
+    }
 
-        for (Bundle bundle : getPackages()) {
-            List<String> jsBundles = getBundleScripts(bundle, appName, key);
-            for (String jsBundle : jsBundles) {
-                if(bundle.getResource(jsBundle) == null) {
-                    logger.error("Application {} declared {} has entry point but file is not found, skipping it from {}",bundle.getSymbolicName(), jsBundle, key);
-                    continue;
+    private void updateBundleInfo(Map<String,AppInfo> newAppInfos, Bundle bundle) {
+        JSONObject pkgJson = null;
+        try {
+            pkgJson = new JSONObject(IOUtils.toString(bundle.getResource(PACKAGE_JSON) != null ? bundle.getResource(PACKAGE_JSON) : bundle.getResource(JAHIA_JSON)));
+            if (pkgJson.has(JAHIA)) {
+                if (pkgJson.getJSONObject(JAHIA).has(URLPATTERNS)) {
+                    processUrlPatterns(newAppInfos, pkgJson);
+                }
+                if (pkgJson.getJSONObject(JAHIA).has(REMOTES)) {
+                    processKeyScripts(newAppInfos, pkgJson, bundle, REMOTES);
                 }
-                if (jsBundle.startsWith("/")) {
-                    resources.add(jsBundle);
-                } else {
-                    resources.add("/modules/" + bundle.getSymbolicName() + "/" + jsBundle);
+                if (pkgJson.getJSONObject(JAHIA).has(APPS)) {
+                    processKeyScripts(newAppInfos, pkgJson, bundle, APPS);
                 }
             }
+        } catch (JSONException | IOException e) {
+            logger.error("Error processing bundle JSON descriptor for bundle {}", bundle.getSymbolicName(), e);
         }
-        return resources;
     }
 
-    private List<Bundle> getPackages() {
-        return Arrays.stream(FrameworkService.getBundleContext().getBundles()).filter(bundle -> bundle.getState() == BundleState.ACTIVE.toInt() && BundleUtils.isJahiaModuleBundle(bundle) && (bundle.getResource(PACKAGE_JSON) != null || bundle.getResource(JAHIA_JSON) != null)).collect(Collectors.toList());
+    interface ValueProcessor {
+        void processValues(List<String> values, AppInfo appInfo);
```

Comment:
Would add a bit of java doc or comment to explain the expected inputs formats and parameters.

Reply:
Ok good point, will do that.